File: project1/src/my_project_cpp/src/counter.cpp
#include "ros/ros.h"

// Messagens
#include "std_msgs/Float64.h"
#include "std_srvs/Empty.h"
#include "my_project_msg/CounterHistory.h"


class Counter{

    public:
        Counter(ros::NodeHandle *nh){
            count = 0;
            publish_interval = 1;

            if (nh->getParam("initial_count", count) ){
                ROS_INFO("Contagem inicial em %f", count);
            }
            else{
                ROS_INFO("Contagem inicial nao definido");
            }   
            nh->param<double>("pub_rate", publish_interval, 1);

            if (nh->hasParam("pub_rate") || nh->hasParam("initial_count")  ){
                ROS_WARN("Parametros customizados");
                nh->setParam("custom_param", true);
            }
            else{
                nh->setParam("custom_param", false);
            }

            if (nh->hasParam("delete_param")){
                ROS_ERROR("Deletando Parametros");
                nh->deleteParam("pub_rate");
                nh->deleteParam("initial_count");
            }

            last_count = count;
            cycles = 0;

            num_sub = nh->subscribe("/number", 10, &Counter::numberCallback, this);
            count_pub = nh->advertise<std_msgs::Float64>("/current_count", 10);

            timer_pub = nh->createTimer(ros::Duration(publish_interval), &Counter::timerCallback, this);
            reset_srv = nh->advertiseService("/reset_counter", &Counter::resetSrvCallback, this);

            history_pub = nh->advertise<my_project_msg::CounterHistory>("/history_counter", 10);
        }

        void numberCallback(const std_msgs::Float64 &msg){
            ROS_INFO("Callback chamado com valor: %f", msg.data);
            cycles++;
            last_count = count;
            count += msg.data;
            ROS_INFO("Contagem Atual %f", count);
        }

        void timerCallback(const ros::TimerEvent &event){
            ROS_INFO("Timer chamado");
            std_msgs::Float64 msg;
            msg.data = count;
            count_pub.publish(msg);

            my_project_msg::CounterHistory history_msg;
            history_msg.current_value = count;
            history_msg.last_value = last_count;
            history_msg.cycles = cycles;
            history_pub.publish(history_msg);
        }

        bool resetSrvCallback(std_srvs::Empty::Request &req, std_srvs::Empty::Response &res){
            count = 0;
            ROS_INFO("Resetando a contagem.");
            return true;
        }

        
    private:
        double count;
        double publish_interval;
        double last_count;
        int cycles;

        ros::Publisher history_pub;
        ros::Subscriber num_sub;
        ros::Publisher count_pub;
        ros::Timer timer_pub;
        ros::ServiceServer reset_srv;
        ros::ServiceServer check_greater_srv;
};


int main(int argc, char **argv){

    ros::init(argc, argv, "counter_node");
    ros::NodeHandle nh;
    Counter counter = Counter(&nh);
    ros::spin();

    return 0;
}

File: project2/catkin_ws/src/my_project_py/scripts/counter.py
#!/usr/bin/env python3
import rospy
from std_msgs.msg import Float64
from std_srvs.srv import Empty, EmptyResponse
from my_project_msg.msg import CounterHistory
from my_project_msg.srv import CheckNumber , CheckNumberResponse

class Counter:

    def __init__(self):
        self.count = rospy.get_param("initial_count", 0)
        self.publish_interval = rospy.get_param("~pub_rate", 3)
        
        
        if(rospy.has_param("~pub_rate") or rospy.has_param("initial_count")):
            rospy.loginfo("parametros customizados")
            rospy.set_param("~custom_param", True)
        else:
            rospy.set_param("~custom_param", False)
            
        if(rospy.has_param("~delete_param")):
            rospy.loginfo("deletando parametros")   
            if(rospy.has_param("~pub_rate")):
                rospy.delete_param("~pub_rate") 
            if(rospy.has_param("initial_count")):
                rospy.delete_param("initial_count")     
        self.last_count = self.count
        self.cycles = 0    
       
        self.num_sub = rospy.Subscriber("/number", Float64, self.numberCallback, queue_size = 10)
        self.count_pub = rospy.Publisher("/current_count", Float64, queue_size = 10)
        self.timer_pub = rospy.Timer(rospy.Duration(self.publish_interval), self.timerCallback)
        self.reset_srv = rospy.Service('reset_counter', Empty, self.resetSrvCallback)
        self.history_pub = rospy.Publisher("/history_count", CounterHistory, queue_size = 10)
        self.check_less__srv = rospy.Service('check_less', CheckNumber, self.checkLessSrvCallback)
        
    def numberCallback(self, msg):
        self.cycles = self.cycles + 1
        self.last_count = self.count
        self.count = self.count + msg.data
        rospy.loginfo("Contagem Atual " + str(self.count))


    def timerCallback(self, event):
        msg = Float64()
        msg.data = self.count
        self.count_pub.publish(msg)
        history_msg = CounterHistory()
        history_msg.current_value = self.count
        history_msg.last_value = self.last_count
        history_msg.cycles = self.cycles
        self.history_pub.publish(history_msg)
        
    def resetSrvCallback(self, req):
        self.count = 0 
        rospy.loginfo("resetando a contagem")
        return EmptyResponse()  

        
    def checkLessSrvCallback(self, req):
        
        rospy.loginfo("Verificando se numero eh maior que a contagem. ")
        res = CheckNumberResponse()
        if(req.number > self.count):
            res.result = True
        else:
            res.result = False    
        return res
if __name__ == '__main__':
    try:
        rospy.init_node("counter_node")
        Counter()
        rospy.spin()
    except rospy.ROSInterruptException:
        pass